File: 345-reverse-vowels/solution.go
package main

import "fmt"

func reverseVowels(s string) string {
	t := []byte(s)
	l, r := 0, len(s)-1
	for l < r {
		for l < r && !isVowel(s[l]) {
			l++
		}
		for l < r && !isVowel(s[r]) {
			r--
		}
		t[l], t[r] = t[r], t[l]
		l++
		r--
	}
	return string(t)
}
func isVowel(c byte) bool {
	switch c {
	case 'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U':
		return true
	default:
		return false
	}
}
func main() {
	s := "IceCreAm"
	fmt.Println("Ans = ", reverseVowels(s))
}

/*
switch case用法:
LanguagePractice/GoPractice/SwitchCase.go

官方解法是用strings.Contains效率可能更好
https://leetcode.cn/problems/reverse-vowels-of-a-string/

!strings.Contains("aeiouAEIOU", string(t[i]))
等價意思是：
「t[i] 這個字元 不在 母音集合 aeiouAEIOU 裡」。

參數對應
第一個 "aeiouAEIOU"：候選集合（你要去哪裡找）
第二個 string(t[i])：目前字元（你要找什麼）

其它 string 方法可以看
LanguagePractice/GoPractice/packageString.go
*/

File: 345-reverse-vowels/Solution.java
class Solution345 {
    public static boolean isVowel(char ch){
        return "aeiouAEIOU".indexOf(ch) < 0;
    }
    public static void swap(char[] arr, int i, int j){
        char temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
    public static String reverseVowels(String s) {
        int l = 0, r = s.length()-1;
        char[] arr = s.toCharArray();
        while (l < r){
            while (l < r && isVowel(s.charAt(l)))
                l++;
            while (l < r && isVowel(s.charAt(r)))
                r--;
            swap(arr, l, r);
            l++;
            r--;
        }
        return new String(arr);
    }
    static void main(String[] args) {
        String s = "IceCreAm";
        System.out.println("Ans = " + reverseVowels(s));
    }
}
/*
indexOf 是用來找「某個字元/子字串第一次出現的位置（索引）」的方法。
indexOf 就是「找位置，找不到回 -1」
 */